File: tree.pl
#!/usr/bin/perl
use strict; use warnings;

sub findRoot{
	my $ID = $_[0];
	$ID =~ m/(\d{16})(\d{16})/;
	my $parent = $1;
	my $child = $2;
	my @names = `ls users`;
	while($1 ne $2){#is not root
		foreach(@names){
			$_ =~ m/(\d{16})(\d{16})/;
			if($2 eq $parent){
				$parent = $1;
				$child = $2;
				$ID = $_;
				last;
			}
		}
		$ID =~ m/(\d{16})(\d{16})/;
	}
	$ID =~ m/(\d{16})(\d{16})/;
	if($1 eq $2){
		$ID;
	}
}

sub buildTree
{
	my $parentID = $_[0]; chomp($parentID);
	my @ls = `ls users/`;
	my $tabs = (defined $_[1])? $_[1] : 0;
	foreach(@ls){
		my $parent = substr($parentID, 16, 16);
		if($_ =~ m/^($parent\d{16})$/ && $1 ne $parentID){#found a son
			my $t = "";
			for(1..$tabs){#append tabs
				$t .= "\t";
			}
			print "$t$1\n";
			&buildTree($1, $tabs+1);
		}
	}
}



&buildTree(&findRoot("13754285045848371375428518689530"));
#print &findRoot("13754285045848371375428518689530");

sub mkdirs
{
	my $number = time;
	chomp($number);
	my $extraNumber = int(rand()*1000000);
	while($extraNumber<100000){############## IN RUN.CGI ÜBERNEHMEN!!!
		$extraNumber = int(rand()*1000000);
	}
	chomp($extraNumber);
	$number .= $extraNumber;
	chomp($number);
	if(defined $_[0]){
		my $pID = substr($_[0], 16, 16);
		`mkdir users/$pID$number`;
	}
	else{
		`mkdir users/$number$number`;
	}
	print $number."\n";
}

#&mkdirs("13754284785691341375428504584837");

File: tree.php
<?php

$idLength = 16;
$pattern = "/([0-9]{".$idLength."})([0-9]{".$idLength."})$/";

function findRoot($ID) {
	global $idLength;
	global $pattern;
	preg_match($pattern, $ID, $matches);
    $parent = $matches[1];
    $child = $matches[2];
	$names = lsToArray("users/*");
	while($matches[1] != $matches[2]){#is not root
		foreach($names as $value){
			preg_match($pattern, $value, $matches);
			if($matches[2] == $parent){
				$parent = $matches[1];
				$child = $matches[2];
				$ID = $value;
				break;
			}
		}
	preg_match($pattern, $ID, $matches);
	}

	preg_match($pattern, $ID, $matches);
	if($matches[1] == $matches[2]){
		return $ID;
	}
}

//echo(findRoot("13794182717410031379418417430622"));
print_r(unserialize(serialize(buildtree(findRoot("13794182717410031379418417430622")))));

function buildTree($parentID) {
	$tree = array();
	global $idLength;
    array_push($tree, $parentID);
    $ls = lsToArray("/home/tobias/Desktop/metap/metaP3/users/*");
    foreach ($ls as $value) {
        $parent = substr($parentID, $idLength, $idLength);
        if (preg_match("/" . $parent . "[0-9]{" . $idLength . "}/", $value, $matches) && $matches[0] != $parentID) {#found a son
			array_push($tree, buildTree($value));
        }
    }
    return $tree;
}

function lsToArray($path) {
    $ls = glob($path, GLOB_ONLYDIR);
    foreach ($ls as &$value) {
        $value = basename($value);
    }
    return $ls;
}
?>

